File: dawn/src/main/java/com/lzq/dawn/util/thread/UtilsThreadFactory.java
package com.lzq.dawn.util.thread;

import android.util.Log;

import androidx.annotation.NonNull;

import java.util.concurrent.ThreadFactory;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicLong;

/**
 * @Name :UtilsThreadFactory
 * @Time :2022/9/2 14:46
 * @Author :  Lzq
 * @Desc :
 */
public class UtilsThreadFactory extends AtomicLong
        implements ThreadFactory {
    private static final AtomicInteger POOL_NUMBER      = new AtomicInteger(1);
    private static final long          serialVersionUID = -9209200509960368598L;
    private final        String        namePrefix;
    private final        int           priority;
    private final        boolean       isDaemon;

    UtilsThreadFactory(String prefix, int priority) {
        this(prefix, priority, false);
    }

    UtilsThreadFactory(String prefix, int priority, boolean isDaemon) {
        namePrefix = prefix + "-pool-" +
                POOL_NUMBER.getAndIncrement() +
                "-thread-";
        this.priority = priority;
        this.isDaemon = isDaemon;
    }

    @Override
    public Thread newThread(@NonNull Runnable r) {
        Thread t = new Thread(r, namePrefix + getAndIncrement()) {
            @Override
            public void run() {
                try {
                    super.run();
                } catch (Throwable t) {
                    Log.e("ThreadUtils", "Request threw uncaught throwable", t);
                }
            }
        };
        t.setDaemon(isDaemon);
        t.setUncaughtExceptionHandler(new Thread.UncaughtExceptionHandler() {
            @Override
            public void uncaughtException(Thread t, Throwable e) {
                System.out.println(e);
            }
        });
        t.setPriority(priority);
        return t;
    }
}

File: dawn/src/main/java/com/lzq/dawn/util/thread/UtilsThreadFactory.kt
package com.lzq.dawn.util.thread

import android.util.Log
import java.util.concurrent.ThreadFactory
import java.util.concurrent.atomic.AtomicInteger
import java.util.concurrent.atomic.AtomicLong

/**
 * @Name :UtilsThreadFactory
 * @Time :2022/9/2 14:46
 * @Author :  Lzq
 * @Desc :
 */
class UtilsThreadFactory @JvmOverloads internal constructor(
    prefix: String,
    private val priority: Int,
    private val isDaemon: Boolean = false
) : AtomicLong(), ThreadFactory {
    private val namePrefix: String

    init {
        namePrefix = prefix + "-pool-" + POOL_NUMBER.getAndIncrement() + "-thread-"
    }

    override fun newThread(r: Runnable): Thread {
        val t: Thread = object : Thread(r, namePrefix + andIncrement) {
            override fun run() {
                try {
                    super.run()
                } catch (t: Throwable) {
                    Log.e("ThreadUtils", "Request threw uncaught throwable", t)
                }
            }
        }
        t.isDaemon = isDaemon
        t.uncaughtExceptionHandler = Thread.UncaughtExceptionHandler { _, e -> println(e) }
        t.priority = priority
        return t
    }

    companion object {
        private val POOL_NUMBER = AtomicInteger(1)
        private const val serialVersionUID = -9209200509960368598L
    }

    override fun toByte(): Byte {
        return super.get().toByte()
    }

    override fun toShort(): Short {
     return super.get().toShort()
    }
}